components/usage/pkg/controller/reconciler.go

```diff
@@ -71,11 +74,59 @@ func (u *UsageReconciler) ReconcileTimeRange(ctx context.Context, from, to time.
 	if len(invalidInstances) > 0 {
 		log.WithField("invalid_workspace_instances", invalidInstances).Errorf("Detected %d invalid instances. These will be skipped in the current run.", len(invalidInstances))
 	}
-
 	log.WithField("workspace_instances", instances).Debug("Successfully loaded workspace instances.")
+
+	workspaces, err := u.loadWorkspaces(ctx, instances)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load workspaces for workspace instances in time range: %w", err)
+	}
+	status.Workspaces = len(workspaces)
+
 	return status, nil
 }
 
+type workspaceWithInstances struct {
+	Workspace db.Workspace
+	Instances []db.WorkspaceInstance
+}
```

Comment:
**Thought**: I wonder if we actually want to load into memory the "fully-hydrated" shapes of both `Workspace` and `WorkspaceInstance`.

In the past, we've had issues with memory consumption, because we were keeping full shapes in memory, while only a few fields were actually relevant. This is especially true for workspace instances, because there are so many of them.

A usage record will probably look something like this:

```ts
export type WorkspaceUsageRecord = {
    instanceId: string;
    workspaceId: string;
    userId: string;
    projectId?: string;
    teamId?: string;
    fromTime: string;
    toTime: string;
};
```

So, in my mind, the most efficient flow to get these usage records is:
1. Query only `instanceId`, `workspaceId`, `creationTime`, `stoppedTime` from `d_b_workspace_instance` (for the current period) -- all the other fields are unnecessary
2. Trim the time bounds between `beginningOfMonth` and `min(getCurrentTime(), endOfMonth)` (also when there is no `stoppedTime` yet)
3. For each unique `workspaceId` query only the owner `userId` and possible `projectId` from `d_b_workspace`
4. For each `userId`, query only the possible `teamId` (randomly pick one team for now)

If we query and then load the full `Workspace` and `WorkspaceInstance` shapes into memory, and start passing them around a lot, I worry that we'll have a bad time (because many workspace instances are created every month, and that number is rapidly increasing over time).

Reply:
Sure, this makes sense. But I don't think it makes sense to do this now.

For me, this is an optimisation once we have something working. For now, I'd prefer to waste memory to get to a working version and then measure what is in fact the cost of these queries.

Reply:
Fair point, thanks. (I don't really see why it's better to ship a less optimized version first when we know how to easily write a more optimized one, but I also agree about shipping this sooner than later, and maybe my suggestion could already be considered premature optimization). So, fine to defer this to later. 👍

Reply:
Largely because I think the current handling will go away anyway once we do add attribution to the WSI. If we optimized here, it would then be wasted work because it would also be replaced by the attribution work from WSI.